src/webgpu/api/validation/queue/writeBuffer.spec.ts

```diff
@@ -11,10 +11,115 @@ Tests writeBuffer validation.
 
 Note: destroyed buffer is tested in destroyed/.
 
-TODO: implement.
+TODO: implement usage flag validation.
 `;
 
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { TypedArrayBufferView, TypedArrayBufferViewConstructor } from '../../../gpu_test.js';
 import { ValidationTest } from '../validation_test.js';
 
 export const g = makeTestGroup(ValidationTest);
+
+g.test('ranges')
+  .desc(
+    `
+Tests that the data ranges given to GPUQueue.writeBuffer() are properly validated. Tests calling
+writeBuffer with both TypedArrays and ArrayBuffers and checks that the data offset and size is
+interpreted correctly for both.
+
+  - When passing a TypedArray the data offset and size is given in elements.
+  - When passing an ArrayBuffer the data offset and size is given in bytes.
+
+Also verifies that the specified data range:
+
+  - Describes a valid range of the source buffer.
+  - Fits fully within the destination buffer.
+  - Is 4-byte aligned.
+`
+  )
+  .fn(async t => {
+    const queue = t.device.defaultQueue;
+
+    function runTest(arrayType: TypedArrayBufferViewConstructor, testBuffer: boolean) {
+      const elementSize = arrayType.BYTES_PER_ELEMENT;
+      const buffer = t.device.createBuffer({
+        size: 16 * elementSize,
+        usage: GPUBufferUsage.COPY_DST,
+      });
+      const arraySm: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(8).buffer
+        : new arrayType(8);
+      const arrayMd: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(16).buffer
+        : new arrayType(16);
+      const arrayLg: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(32).buffer
+        : new arrayType(32);
+
+      if (testBuffer) {
+        const array15 = new Uint8Array(15).buffer;
+
+        // Writing the full buffer that isn't 4-byte aligned.
+        t.shouldThrow('OperationError', () => queue.writeBuffer(buffer, 0, array15));
```

Comment:
For negative tests to be as strong as possible, I think this should be "minimally unaligned", i.e. a multiple of the biggest power of 2 that's still invalid. That is, I think testing "14 (multiple of 2) is invalid" is stronger than testing "15 (multiple of 1) is invalid". Similar in other places in this test.

Reply:
I don't think there's anything to read up on, I kind of made it up, but basically: Alignments are generally to powers of two. If I call 8 "more aligned" than 4, 4 "more aligned" than 2, etc., and we want to test two things:
- The implementation doesn't require a bigger alignment (say, 8 instead of 4) than the spec.
- The implementation doesn't require a smaller alignment (say, 2 instead of 4) than the spec.

Thus for positive cases you want to test "least aligned" valid values, (12, say, to check the impl requires a multiple of 4 or smaller)
while for negative cases you want to test "most aligned" invalid values (10, say, to check the impl requires a multiple of 4 or bigger)